### PR TITLE
BUGFIX: TransientNodeCache must be able to return null

### PR DESCRIPTION
**Pull request by kitsunet**
This is an important bugfix because right now non existing nodes will result in a fatal error 
due to the type hint, when the $getter() actually doesn't resolve a node but null, which can happen.
The current behavior is therefore broken for many installations.

Relate #2301

### CODE REVIEW

**Review by kdambekalns — APPROVED** [review]

**Review by bwaidelich — CHANGES_REQUESTED** [review]
Please wait a minute

**Comment by bwaidelich** [comment]
Woa, `TransientNodeCache`, this is the first time that I see this one o.O

But I just realized why. It was introduced 2 years ago (with a38353be4bc676ce5578815c687ea4c9554007fb) but only merged a few days ago.

Instead of fixing this can we please simplify this (I can take care if you agree)?

We all hate runtime caches but need them from time to time. But a runtime cache that injected??

I might be missing the point but currently it's only used in one place currently, so why is:

```php
<?php
class NodePrivilegeContext
{

    /**
     * @Flow\Inject
     * @var TransientNodeCache
     */
    protected $transientNodeCache;

    protected function getNodeByIdentifier($nodeIdentifier)
    {
        return $this->transientNodeCache->cache($nodeIdentifier, function () use ($nodeIdentifier) {
            $context = $this->contextFactory->create([
                // as we are often in backend, we should take invisible nodes into account properly when resolving Node Identifiers to paths.
                'invisibleContentShown' => true
            ]);
            $node = null;
            $this->securityContext->withoutAuthorizationChecks(function () use ($nodeIdentifier, $context, &$node) {
                $node = $context->getNodeByIdentifier($nodeIdentifier);
            });
            $context->getFirstLevelNodeCache()->setByIdentifier($nodeIdentifier, null);
            return $node;
        });
    }
}

class TransientNodeCache
{

    private $nodes = [];

    public function cache(string $nodeIdentifier, callable $getter): NodeInterface
    {
        if (array_key_exists($nodeIdentifier, $this->nodes)) {
            return $this->nodes[$nodeIdentifier];
        }

        $node = $getter();
        if ($node instanceof NodeInterface) {
            $this->nodes[$nodeIdentifier] = $node;
        }
        return $node;
    }
}
```

any better than

```php
<?php
class NodePrivilegeContext
{

    /**
     * @var NodeInterface[]
     */
    private $nodeRuntimeCache = [];

    protected function getNodeByIdentifier($nodeIdentifier)
    {
        if (!array_key_exists($nodeIdentifier, $this->nodeRuntimeCache)) {
            $context = $this->contextFactory->create([
                // as we are often in backend, we should take invisible nodes into account properly when resolving Node Identifiers to paths.
                'invisibleContentShown' => true
            ]);
            $node = null;
            $this->securityContext->withoutAuthorizationChecks(function () use ($nodeIdentifier, $context, &$node) {
                $node = $context->getNodeByIdentifier($nodeIdentifier);
            });
            $context->getFirstLevelNodeCache()->setByIdentifier($nodeIdentifier, null);
            return $node;
        }
        return $this->nodeRuntimeCache[$nodeIdentifier];
    }
}
```

**Comment by kdambekalns** [comment]
> I might be missing the point but currently it's only used in one place currently, so why is: […] any better than […]

Maybe @j6s remembers why he changed that?

**Comment by bwaidelich** [comment]
> fix it to use static access and actually write nodes to the cache property…

aye sir :)

I just inlined his original code to make my point, but you're right of course.
I'll prepare a separate PR and then we can decide once we have feedback from Johannes

**Comment by kitsunet** [comment]
Might have been planned to be used for other privileges as well? I for one jusut wanted to get this nasty bug fixed fast. And I would suggest we merge the fix now and refactor with a bit more relaxed timeline?

**Comment by bwaidelich** [comment]
> Might have been planned to be used for other privileges as well?

Yes, that's possible. But I would suggest that we reconsider once we really need it in multiple places (taking into account that other places might not need the same content context etc)

> I would suggest we merge the fix now and refactor with a bit more relaxed timeline?

Here is my suggestion: #3098

But the choice is yours of course!

**Comment by j6s** [comment]
I am fuzzy on the details of this - but IIRC I initially added it as a simple property on the class, but we came up with the `TransientNodeCache` in code review to make it more explicit what's going on.

I'll look for the original PR when I have a couple of minutes.

**Comment by bwaidelich** [comment]
Thanks for the update, I just read the comments on the original PR (https://github.com/neos/neos-development-collection/pull/2302#issuecomment-444091719) and understand the reasoning behind this.

I still find it a bit over-complex and would prefer a private runtime cache (potentially with the possibility to flush it if we really need it for tests). But with all the context I'm fine with both solutions

**Comment by kdambekalns** [comment]
The let's fix it! Improvement (potentially) in #3098.

**Comment by bwaidelich** [comment]
Hi all,

after sleeping on it I do have a follow-up question.

**TL;DR:** Can't we just fix the issue by removing the `$context->getFirstLevelNodeCache()->setByIdentifier($nodeIdentifier, null);` line?

We currently do (schematic):

```
get Node from the new TransientNodeCache
OR get the Node from the Context's transient node cache (FirstLevelNodeCache)
OR load it from the DB

and then remove the node from the Context cache again
```

This is all required because the `FirstLevelNodeCache` caches the nodes independently from the access level (i.e. `invisibleContentShown` etc)…

Apparently. Because in fact the `FirstLevelNodeCache` is instantiated in the `ContentContext` constructor so it is bound to that instance. And by using the `ContextFactory` that same instance is reused depending on its properties.

=> just removing the line `$context->getFirstLevelNodeCache()->setByIdentifier($nodeIdentifier, null);` should have fixed the issue or what am I missing?

**Comment by kitsunet** [comment]
I assumed!! that there might be instances were there is actually null returned from the DB as well. But _maybe_ you are right. I have to look at it again with a clear head.

**Comment by bwaidelich** [comment]
> there might be instances were there is actually null returned from the DB as well

Sure, but that should not be an issue. The `null` value is also put in the `FirstLevelNodeCache`: https://github.com/neos/neos-development-collection/blob/4.3/Neos.ContentRepository/Classes/Domain/Service/Context.php#L274-L279

and `NodePrivilegeContext::getNodeByIdentifier()` will just return that `null` like today.

**Comment by kdambekalns** [comment]
> Sure, but that should not be an issue.

It is. The return value of `cache()` must be nullable then, which it wasn't…

**Comment by bwaidelich** [comment]
> The return value of cache() must be nullable then, which it wasn't…

`cache()` would not exist any more.
We currently have two nested transient caches, but I claim that we don't need that, so `TransientNodeCache` would go
